tt_metal/tt_stl/any_range.hpp

```diff
@@ -0,0 +1,887 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <algorithm>
+#include <iterator>
+#include <type_traits>
+#include <utility>
+
+namespace tt::stl {
```

Comment:
Can you please add a comment describing what AnyRange is, and how to use it?

Can you please also clarify, what's the difference between `tt::stl::AnyRange` and `boost::any_range`

Reply:
`boost::any_range` does not guarantee that type-erasure is free of dynamic allocation. It has short buffer optimization, but if the container's iterators are too large, it implicitly falls back to using dynamic allocation for storage. This enforces at compile-time that only containers which fit can be type-erased without dynamic allocation, and unlike `boost::any_range` which only supports non-owning type-erasure, this supports both.

Reply:
Sounds good, let's add the comment in the code describing this